Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
-    return ret;
-}
+    // Protects _dispatchQueue, _underlyingQueue, and _qualityOfService
+    // Should never be locked before _concurrentOperationCountLock
+    std::recursive_mutex _dispatchQueueLock;
 
-void NSAtomicListReverse(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* cur = *listPtr;
-    struct NSAtomicListNode* prev = NULL;
-    struct NSAtomicListNode* next = NULL;
+    StrongId<NSMutableArray> _operations;
 
-    if (!cur) {
-        return;
-    }
+    std::atomic_bool _suspended;
 
-    do {
-        next = cur->next;
-        cur->next = prev;
+    // Tracks the number of currently executing operations, for use with maxConcurrentOperationCount
+    NSInteger _currentConcurrentOperationCount;
+    NSInteger _maxConcurrentOperationCount;
 
-        if (next) {
-            prev = cur;
-            cur = next;
-        }
-    } while (next);
+    // Enqueued operations to execute once execution is possible
+    // Enqueued elements are:
+    // 1) already stored in _operations, so don't need to be separately retained
+    // 2) delayed due to the queue's suspension, or due to hitting maxConcurrentOperations
+    // Non-ready operations are handled separately
+    std::queue<NSOperation*> _queuedOperations;
+    std::recursive_mutex _concurrentOperationCountLock; // Protects _currentConcurrentOperationCount and _queuedOperations
+}
 
-    *listPtr = cur;
+// Private helper that checks whether maxConcurrentOperationCount has been exceeded
+- (BOOL)_belowMaxConcurrentOperations {
+    std::lock_guard<std::recursive_mutex> lock(_concurrentOperationCountLock);
+    NSInteger maxOps = self.maxConcurrentOperationCount;
+    return (maxOps <= 0) || (_currentConcurrentOperationCount < maxOps);
 }
 
-NSObject* NSAtomicListPop(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* node = *listPtr;
-    if (!node) {
-        return NULL;
+// Private helper that tries to starts a single NSOperation on the dispatch queue and manages associated state.
+// If the operation is not startable due to the state of the queue (ie: suspension, above max concurrent operations)
+// it is enqueued for later.
+// If the operation is not ready, this is a no-op
+// (the operation can try to start again later once the queue receives a KVO notification that it became ready)
+- (void)_startOperation:(NSOperation*)operation {
+    if (operation.isExecuting || operation.isFinished) {
+        TraceError(g_logTag, L"Attempted to start an operation that was already executing or finished; ignoring.");
+        return;
     }
 
-    *listPtr = node->next;
+    if (!operation.isReady) {
+        // Deliberate no-op (see function comments)
+        return;
+    }
 
-    NSObject* elt = [[node->elt retain] autorelease];
-    delete node;
-    return elt;
-}
+    {
+        std::lock_guard<std::recursive_mutex> lock(_concurrentOperationCountLock);
+        if ((self.isSuspended) || (![self _belowMaxConcurrentOperations])) {
```

Comment:
when we add the property for suspended, this should become `self.suspended`; the getter form is for `[self isSuspended]` use